### ISSUE TITLE
agreeR syntax to use

### ISSUE THREAD

**Issue by troettge** [learnB4SS/learnB4SS#39]
We probably should all agree on and stick to one and the same syntax for the same thing (to avoid confusion).

Here are some things that come up often:
- define priors (with or without quotation marks etc.)
- arguments of brm (should probably always define the same arguments across applications)
- extract posteriors
- plot extracted posteriors

I think participants would benefit from seeing occurring syntax for these things (feel free to add stuff). You guys are more up to date with code-efficient ways to extract posteriors and plotting, so feel free to suggest it here.

**Comment by jvcasillas** [comment]
This all seems reasonable to me. Short chunks over long ones, consistent syntax. I generally use a similar approach to modeling as the one stef shows above (but I keep the outcome variable and pop. effects on the same line). I plan on keep all plots quite simple. They usually go something like this for me: 

```
data %>% 
  ggplot(., aes(x = x, y = y, color = whatever)) + 
    geom_something() + 
    my_theme()
```

... where I use 2 spaces of indentation from data to the start of the plot and then 2 more spaces for each layer added. I also oftern keep the ".", even though it is not necessary. In my experience this has been helpful for students (particularly because of cases where the "data" argument isn't first), but I am pretty flexible on all of this.

**Comment by troettge** [comment]
@stefanocoretta @jvcasillas what about posterior extraction and then plotting / calculating summary stats on it? preferences how to best do it?

**Comment by stefanocoretta** [comment]
I usually just use `posterior_sample()` to extract posteriors but sometimes I use `tidybayes::spread/gather_draws()`. No preference.

**Comment by troettge** [comment]
@jvcasillas @stefanocoretta please add a full code example for your preferred way of posterior extraction and plotting for reference!

**Comment by troettge** [comment]
@jvcasillas @stefanocoretta maybe we can also compile a list of priors that we will be using here?

**Comment by troettge** [comment]
> This all seems reasonable to me. Short chunks over long ones, consistent syntax. I generally use a similar approach to modeling as the one stef shows above (but I keep the outcome variable and pop. effects on the same line). I plan on keep all plots quite simple. They usually go something like this for me:
> 
> ```
> data %>% 
>   ggplot(., aes(x = x, y = y, color = whatever)) + 
>     geom_something() + 
>     my_theme()
> ```
> 
> ... where I use 2 spaces of indentation from data to the start of the plot and then 2 more spaces for each layer added. I also oftern keep the ".", even though it is not necessary. In my experience this has been helpful for students (particularly because of cases where the "data" argument isn't first), but I am pretty flexible on all of this.

It would be good if we had a concrete code example and all use the same one. Like what `geom_` are we using (density?) and how do we arrive at `data`. If some of you has prepared parts of the exercises already, can you share your rmarkdown or is it up already somewhere?

**Comment by jvcasillas** [comment]
Posterior extraction: 

```
posterior_samples(mod)
```

Plotting examples: (will update after finishing 06 slides)

**Comment by jvcasillas** [comment]
> > This all seems reasonable to me. Short chunks over long ones, consistent syntax. I generally use a similar approach to modeling as the one stef shows above (but I keep the outcome variable and pop. effects on the same line). I plan on keep all plots quite simple. They usually go something like this for me:
> > ```
> > data %>% 
> >   ggplot(., aes(x = x, y = y, color = whatever)) + 
> >     geom_something() + 
> >     my_theme()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ... where I use 2 spaces of indentation from data to the start of the plot and then 2 more spaces for each layer added. I also oftern keep the ".", even though it is not necessary. In my experience this has been helpful for students (particularly because of cases where the "data" argument isn't first), but I am pretty flexible on all of this.
> 
> It would be good if we had a concrete code example and all use the same one. Like what `geom_` are we using (density?) and how do we arrive at `data`. If some of you has prepared parts of the exercises already, can you share your rmarkdown or is it up already somewhere?

Ok. I will go back to the first set of slides and be more specific about what I name things in the walkthrough and we can maintain consistency throughout. For example, the first model (intercept only) can be something simple like `b_mod_00` where `b` represents "Bayesian", `mod` because it is a model, and `00` represents a sequential numbering of the model build (00 because it's the first). This will allow us to build all models and know how to name theme (though we may have to renumber once everything is finished and we have final counts. 

We can use similar conventions for posterior data frames and plots. To get the conversation started I will suggest `post_00` for "posterior" and the number is equivalent to the numbers from the model object name, and for plots `p_00`, where `p` represents "plot" and the number is also equivalent to the model object number. Thus, we would end up with the following object from our intercept only model: 

```
# Fit model
b_mod_00 <- brm(articulation_rate ~ 1, data = polite)

# Extract posterior
post_00 <- posterior_samples(b_mod_00)

# Plot posterior of b_Intercept
p_00 <- post_00 %>%
  ggplot(., aes(x = b_Intercept)) + 
    stat_halfeye() 
```

**Comment by troettge** [comment]
@joseph but this is only a very limited case. We will often want to plot several coefficient distributions from one model in one plot, could do it with facets, could do it with color and overlapping distributions, but for that we would need something like: 

`post_00 <- posterior_samples(simple_model) %>% 
  select(b_Intercept, b_attitudepol) %>% 
  pivot_longer(cols = c(b_Intercept, b_attitudepol), 
               names_to = "coef", values_to = "posterior" )`

**Comment by troettge** [comment]
Another issue is, people will DEMAND seeing the posteriors for individual levels (which is also much closer to what we want to show people in our papers). This will blow up code. Can we chat about this soon, I really want to get this going

```
# Extract posterior coefficient
post_00 <- posterior_samples(simple_model) %>% 
  select(b_Intercept, b_attitudepol) %>% 
  mutate(polite = b_Intercept + b_attitudepol) %>% 
  rename(informal = b_Intercept) %>% 
  select(-b_attitudepol) %>% 
  pivot_longer(cols = c(polite, informal), 
               names_to = "coef", values_to = "posterior" )


# Wrangle
# Plot posterior of b_Intercept
p_00 <- post_00 %>%
  ggplot(., aes(x = posterior, group = coef,fill = coef)) + 
           stat_halfeye(alpha = 0.5) 

p_00
´´´

**Comment by stefanocoretta** [comment]
http://mjskay.github.io/tidybayes/articles/tidy-brms.html#posterior-fits-1

**Comment by troettge** [comment]
Does not make it much better: Problem not solved:

```
# Extract posterior coefficient
post_01 <- multiple_model %>%
  spread_draws(b_Intercept, b_attitudepol, b_tasknot) %>%
  mutate(polite_dct = b_Intercept + b_attitudepol,
         polite_not = b_Intercept + b_attitudepol + b_tasknot,
         informal_dct = b_Intercept,
         informal_not = b_Intercept + b_tasknot) %>% 
  select(polite_dct, polite_not, informal_dct, informal_not) %>%
  # make tidy long format
  pivot_longer(cols = c(polite_dct, polite_not, informal_dct, informal_not), 
               names_to = "coef", values_to = "posterior" ) %>% 
  mutate(attitude = ifelse(coef %in% c("polite_dct", "polite_not"), "polite", "informal"),
         task = ifelse(coef %in% c("polite_dct", "informal_dct"), "dct", "not"))
  
# Plot posteriors
p_01 <- post_01 %>%
  ggplot(., aes(x = posterior, fill = attitude)) + 
           stat_halfeye(alpha = 0.5) +
  facet_grid(task~.)

p_01
```

**Comment by jvcasillas** [comment]
General comment... I didn't understand exactly what you were concerned about initially because my examples haven't dealt with a model with a categorical variable yet. What you describe above is a "problem" specific to knowing R (general data tidying) and one that arises specifically when dealing with a categorical factor with more than two levels. As we've said from the beginning, this workshop isn't about teaching R. I think there is a concrete solution for this specific use-case. 1) Mention that using  categorical variables with multiple levels creates a general data tidying problem, 2) offer an example solution (mutate, pivot_longer) but leave it up to them to worry about it. This might be unsatisfying as a solution, but stays within the scope of what we want to do. If we want we could simplify and avoid the problem altogether if we a) just use a variable with 2 levels, b) pick another variable, or 3) subset the attitude (or task) variable, etc.

**Comment by jvcasillas** [comment]
Regarding plotting the posteriors of a model 'as-is' (i.e., without adjusting for specific comparisons as in the example above), there are simple one line solutions to generate forest plots in the packages that will be installed as dependencies with our `learnB4SS` package.

**Comment by jvcasillas** [comment]
Also, @stefanocoretta @troettge , In my first presentation I assign the dataset to an object called `polite`, i.e., `polite <- learnB4SS::dataset` and then run examples on the dataset using the `polite` object.  It would be a good idea to keep this consistent too. If you prefer some other name, just let me know and I can change to something else.

**Comment by stefanocoretta** [comment]
I now called it `b4ss_data` but can change to polite which has less chars and only letters.

**Comment by jvcasillas** [comment]
Ah that might be an even easier solution ("after loading the package you will have access to a dataset called 'polite' yadda yadda". Sounds like a plan.

**Comment by stefanocoretta** [comment]
> Ah that might be an even easier solution ("after loading the package you will have access to a dataset called 'polite' yadda yadda". Sounds like a plan.

Done!

**Comment by stefanocoretta** [comment]
Two liner to extract samples.

```r
library(modelr)
library(tidybayes)
library(learnB4SS)
data("polite")

# load b_mod_01 from slides/assets

# Get samples
polite %>%
  data_grid(attitude) %>%
  add_predicted_draws(b_mod_01) %>%
# Now we can plot
  ggplot(aes(x = .prediction, y = attitude)) +
    stat_slab()
```
